### PR TITLE
Fix masked attention with gradient checkpointing

### LINKED ISSUE #2049: MaskedCausalVisionTransformer ignores masks with gradient checkpointing
`MaskedCausalVisionTransformer.forward_features()` ignores its attention mask when gradient checkpointing is enabled. The checkpointed masked output matches an unmasked forward instead of the ordinary masked path.

```python
import torch

from lightly.models.modules.masked_causal_vision_transformer import MaskedCausalVisionTransformer

torch.manual_seed(0)
model = MaskedCausalVisionTransformer(
    img_size=32, patch_size=16, embed_dim=24, depth=2, num_heads=3
).eval()
images = torch.rand(2, 3, 32, 32)
mask = torch.zeros(2, 5, dtype=torch.bool)
mask[:, 1:] = True

expected = model.forward_features(images, mask=mask)
model.set_grad_checkpointing()
actual = model.forward_features(images, mask=mask)

torch.testing.assert_close(actual, expected)
```

Environment: Linux 6.8, Python 3.12.3, PyTorch 2.13.0 CPU.

### PR DESCRIPTION
**Pull request by Purin1410**
Fixes #2049

Forward the attention mask through checkpointed transformer blocks. Add regression coverage for forward and input-gradient parity.

Fix masked attention when gradient checkpointing is enabled. Forward the mask through checkpointed transformer blocks and add regression coverage for output and input-gradient parity.

### CODE REVIEW

**Comment by gabrielfruet** [comment]
Thank you for the PR! these fixes are really great

Do you mind driving the CI to green? should be simple

**Comment by Purin1410** [comment]
Thanks! I rebased this branch onto the current `master` and fixed the type-check failure by narrowing `images.grad` before cloning it and zeroing the existing gradient before the checkpointed backward pass.

Local validation is green:
- `make lock-check`
- `make type-check` (229 source files)
- `make format-check` (506 files)
- focused test module: 6 passed
- `tests/models/modules`: 94 passed, 92 skipped

The four new fork workflow runs are currently marked `action_required` and appear to be waiting for maintainer approval.

**Review by gabrielfruet — APPROVED** [review]
